src/provider.c

```diff
@@ -1045,12 +1045,20 @@ static int p11prov_module_init(P11PROV_CTX *ctx)
         .pReserved = (void *)ctx->init_args,
     };
     CK_INFO ck_info = { 0 };
+    const char *env_module;
     int ret;
 
     if (ctx->status != P11PROV_UNINITIALIZED) {
         return 0;
     }
 
+    /* The environment variable has the highest precedence */
+    env_module = getenv("PKCS11_PROVIDER_MODULE");
+    if (env_module && *env_module) {
+        free(ctx->module);
+        ctx->module = strdup(env_module);
```

Comment:
please use OPENSSL_strdup() and add an OPENSSL_free() of the module parameter in the providered cleanup function.

Reply:
Thanks. Changed.